### ISSUE TITLE
Auto creation of database does not work

### ISSUE THREAD

**Issue by JesperAxelsson** [nurpax/snaplet-sqlite-simple#12]
Hi!
I tried using your snaplet and it doesn't seem to copy the cfg files correctly on windows. The folders is created but no files created. Tried it with your example app as well but no lock. 

Am I  missing a config file or something else?

This were installed in a sandbox I should note.

Error:
Initializer threw an exception...
Must specify db filename

**Comment by nurpax** [comment]
Strange.  I haven't ever run any of this on Windows but I don't see why it still wouldn't work.

Are any of the other Snap samples working for you, like the default sample project?  I think some of those are creating cfg files too.

You could also try creating the cfg files yourself and then starting the server.  For example, here's a couple of examples from my own "weight lifting" web app:

https://github.com/nurpax/hswtrack/blob/master/snaplets/sqlite-auth/test.cfg
https://github.com/nurpax/hswtrack/blob/master/snaplets/sqlite-simple/test.cfg

You can try creating the `snaplets/sqlite-auth` and `snaplets/sqlite-simple` and create cfg files under there.

**Comment by JesperAxelsson** [comment]
Yeah it is. I had the same problem with postgres-simple as well so it might not be your problem actually. It might have something to do with the sandbox, but when I looked at the path with
```
paths <- getSnapletFilePath
liftIO $ putStrLn ("Paths: " ++ paths)
```
the correct path were shown.

I tried to put the cfg files in the same paths as you, but my app can't find for some reason. I'm not clever enough in haskell to understand what is going on in the snaplet lib.

This code were missing from my appInit, is this needed to initiate  snaplet auth?
```
 let c = sqliteConn $ d ^# snapletValue
liftIO $ withMVar c $ \conn -> Model.createTables conn
```

Slick weightlifting app btw :)

Thanks,
Jesper Axelsson

**Comment by nurpax** [comment]
```
let c = sqliteConn $ d ^# snapletValue
liftIO $ withMVar c $ \conn -> Model.createTables conn
```

This is needed in order to create the database but it's not really required for snaplet init.  That's just if you want to do some db initialization on app load.  The error  "Initializer threw an exception: Must specify db filename" is orthogonal to this init code.

It could be the config file can't be found or it somehow fails to parse.  I have a couple of suggestions on how you could debug this further:

 1. Try running the same code on Linux (e.g. create a Virtualbox VM for say Ubuntu)
 2. Try adding some debug prints in the snaplet-sqlite-simple init code in https://github.com/nurpax/snaplet-sqlite-simple/blob/master/src/Snap/Snaplet/SqliteSimple.hs

The snaplet init code is as follows:

```
------------------------------------------------------------------------------
-- | Initialize the snaplet
sqliteInit :: SnapletInit b Sqlite
sqliteInit = makeSnaplet "sqlite-simple" description datadir $ do
    config <- getSnapletUserConfig
    (mci,errs) <- runWriterT $
        logErr "Must specify db filename" $ C.lookup config "db"
    let ci = fromMaybe (error $ intercalate "\n" errs) mci
    tracing <- liftIO $ C.lookupDefault False config "enableSqlTracing"
    conn <- liftIO (S.open ci >>= setTracing tracing >>= newMVar)
    return $ Sqlite conn
  where
    description = "Sqlite abstraction"

    datadir = Just $ liftM (++"/resources/db") getDataDir

    setTracing tracing conn = do
      when tracing (S.setTrace conn (Just (putStrLn . T.unpack)))
      return conn
```

Perhaps print the values of

**Comment by nurpax** [comment]
Also, I'm not sure which directory you use to run your app.

Perhaps try

```
cabal install
cabal run
# try also `cabal run -e test`
```

**Comment by JesperAxelsson** [comment]
Hi!

I finally got a Ubuntu vm running. Your example app for snaplet-sqlite-simple  does work on ubuntu, unfortunatly not on my windows 7 machine. 
My own app is not working on neither ubuntu nor windows, so there is probably a bug there as well. I am not quite sure how to proceed though.
I tried printing the path with `getSnapletFilePath` which does point to the directory with my cabal file.
[My site file if you want to take a look.](https://bitbucket.org/SeriousJope/thefubar/src/7f270cb8a61b925983c15a76ce13e9789e17bd18/src/Site.hs?at=default#cl-334)



Thanks for your help so far!
Jesper Axelsson

**Comment by nurpax** [comment]
Hi @JesperAxelsson sorry, I haven't had time to look into this.   I don't develop Haskell on Windows so it's hard to repro this.  Looks like another person in https://github.com/nurpax/snaplet-sqlite-simple/issues/14 is using Windows and snaplet-sqlite-simple, I guess this code has worked for him. 

@wochinge snaplet-sqlite-simple is working for you on Windows?  Your screenshot looked like you might be running Windows. :)

**Comment by wochinge** [comment]
Yes, it's working like a charm on Windows 8.1 x64, Ubuntu 14.04 and Mac OS!
devel.cfg is in <dir where cabal file is>/snaplets/sqlite-simple/devel.cfg.
Is the structure of your devel.cfg correct and did you check the encoding?

**Comment by JesperAxelsson** [comment]
Thanks! I moved to postgres as was my my actual goal. I wanted to use
sqlite to quickly create a MVP and then move to a bigger database at my
leasure.
One of the things I did wrong were that I accidentally had change the name
of the cfg that snap were looking for.
I could not get it to copy the devel.cfg on my Win7 machine for some
reason. If you want I could test it some more when I have the time?
